route-rs-packets/src/arp.rs

```diff
@@ -0,0 +1,253 @@
+        let (start, end) = self.sender_hardware_addr_range();
+        self.arp_data(start, end)
+    }
+
+    pub fn sender_protocol_addr(&self) -> &[u8] {
+        let (start, end) = self.sender_protocol_addr_range();
+        self.arp_data(start, end)
+    }
+
+    pub fn target_hardware_addr(&self) -> &[u8] {
+        let (start, end) = self.target_hardware_addr_range();
+        self.arp_data(start, end)
+    }
+
+    pub fn target_protocol_addr(&self) -> &[u8] {
+        let (start, end) = self.target_protocol_addr_range();
+        self.arp_data(start, end)
+    }
+
+    pub fn set_hardware_type(&mut self, htype: u16) {
+        let (start, end) = HARDWARE_TYPE_RANGE;
+        self.set_arp_data(&htype.to_be_bytes(), start, end);
+    }
+
+    pub fn set_protocol_type(&mut self, ptype: u16) {
+        let (start, end) = PROTOCOL_TYPE_RANGE;
+        self.set_arp_data(&ptype.to_be_bytes(), start, end);
+    }
+
+    pub fn set_hardware_addr_len(&mut self, len: u8) {
+        let (start, end) = HARDWARE_ADDR_LEN_RANGE;
+        self.set_arp_data(&len.to_be_bytes(), start, end);
+    }
+
+    pub fn set_protocol_addr_len(&mut self, len: u8) {
+        let (start, end) = PROTOCOL_ADDR_LEN_RANGE;
+        self.set_arp_data(&len.to_be_bytes(), start, end);
+    }
+
+    pub fn set_opcode(&mut self, code: u16) {
+        let (start, end) = OPCODE_RANGE;
+        self.set_arp_data(&code.to_be_bytes(), start, end);
+    }
+
+    pub fn set_sender_hardware_addr(&mut self, addr: MacAddr) {
+        // NOTE: should we set len based on frame data, or param?
+        let (start, end) = self.sender_hardware_addr_range();
+        self.set_arp_data(&addr.bytes, start, end);
+    }
+
+    pub fn set_sender_protocol_addr(&mut self, ip_addr: IpAddr) {
+        // NOTE: should we set len based on frame data, or param?
+        let (start, _) = self.sender_protocol_addr_range();
+        self.set_ip_addr(ip_addr, start);
+    }
+
+    pub fn set_target_hardware_addr(&mut self, addr: MacAddr) {
+        // NOTE: should we set len based on frame data, or param?
+        let (start, end) = self.target_hardware_addr_range();
+        self.set_arp_data(&addr.bytes, start, end);
+    }
+
+    pub fn set_target_protocol_addr(&mut self, ip_addr: IpAddr) {
+        // NOTE: should we set len based on frame data, or param?
+        let (start, _) = self.target_protocol_addr_range();
+        self.set_ip_addr(ip_addr, start);
+    }
+
+    // Move ownership of the frame back to the caller
+    pub fn frame(self) -> EthernetFrame {
+        self.frame
+    }
+
+    // Returns the bytes in the ethernet frame between start and end, exclusive
+    fn arp_data(&self, start: usize, end: usize) -> &[u8] {
+        let frame_offset_start = self.frame.payload_offset + start;
+        let frame_offset_end = self.frame.payload_offset + end;
+
+        // TODO: I'd like to use `self.frame.payload()` here, but having ownership difficulties with Cow
```

Comment:
Yeah, this is the battle that you don't "own" the vector, you "own" the EthernetFrame.  So instead of containing the EthernetFrame, just strip the PacketData out into you structure, and set the layer2_offset appropriately.

Reply:
Yeah, I see the lifetime requirements now. If I own the packet vector (as a field in the `ArpFrame` struct) we know any references that are passed out here will live through the caller (at least until `ArpFrame` is `Drop`ped?).

Reply:
I guess we're still using `EthernetFrame` as a client effectively here, it just seems like exposing too much of its internals. As `ArpFrame`, I shouldn't have to know about the internal structure of `EthernetFrame` in order to obtain a reference to its payload.

Reply:
This is more of a nit point I'd like to make people aware of than an urgent problem.

Reply:
Hmm, the PacketData field is public, so as an owner of EthernetFrame you are able to directly change its vec, you just do so at your own peril. 
Payload is supposed to a be a take it and go kind of function.  Or you can wholesale take the payload, modify it, and drop it back into set_payload().  Does cause an allocation though.

Reply:
Exactly! I'd like to only get/set particular ranges of bytes in the underlying vec, but I'm finding it difficult to do that with `EthernetFrame`'s interface atm. Again, not the biggest deal since we're able to reference its `PacketData` here. I'm also able to modify it in the `set_arp_data` fn.

Reply:
Okay, using the `PacketData` escape hatch is fine for now.  Rust's safety might be getting in the way of giving us the interface we would like trivially.